### PR TITLE
SNOW-1729244 support for large and small timestamps

### PR DESCRIPTION
**Pull request by sfc-gh-dstempniak**
### Description
Support for large and small timestamps such as 0001-01-01 and 9999-12-31.

### Checklist
- [x] Code compiles correctly
- [x] Code is formatted according to [Coding Conventions](../blob/master/CodingConventions.md)
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing (`dotnet test`)
- [x] Extended the README / documentation, if necessary
- [x] Provide JIRA issue id (if possible) or GitHub issue id in PR name

### CODE REVIEW

File: Snowflake.Data.Tests/UnitTests/SFDataConverterTest.cs [hunk 1]
```diff
@@ -326,5 +328,25 @@ public void TestInvalidConversionInvalidDecimal(string s)
...
+        {
+            object unsupportedObject;
+            if (unsupportedType == typeof(DateTimeOffset))
+                unsupportedObject = new DateTimeOffset();
+            else if (unsupportedType == typeof(DateTime))
+                unsupportedObject = new DateTime();
+            else
+                unsupportedObject = null;
+
+            Assert.NotNull(unsupportedType);
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data.Tests/UnitTests/SFDataConverterTest.cs:347]: should it by unsupportedType or unsupportedObject?

File: Snowflake.Data.Tests/UnitTests/SFDataConverterTest.cs [hunk 2]
```diff
@@ -326,5 +328,25 @@ public void TestInvalidConversionInvalidDecimal(string s)
...
+            SnowflakeDbException ex = Assert.Throws<SnowflakeDbException>(() => SFDataConverter.csharpValToSfVal(dataType, unsupportedObject));
+            SnowflakeDbExceptionAssert.HasErrorCode(ex, SFError.INVALID_DATA_CONVERSION);
+        }
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data.Tests/UnitTests/SFDataConverterTest.cs:350]: do we need to add anything to:
- StructuredTypesTest.TestTimeConversions
- StructuredTypesWithEmbeddedUnstructuredIT.TestSelectDateTime
- StructuredTypesWithEmbeddedUnstructuredIT.TestSelectDateTimeOffset
?

File: Snowflake.Data.Tests/IntegrationTests/SFBindTestIT.cs [hunk 3]
```diff
@@ -1,6 +1,7 @@
 ﻿/*
  * Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
  */
+#nullable enable
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data.Tests/IntegrationTests/SFBindTestIT.cs:4]: shouldn't we do `#nullable disable` at the end of the file?

File: Snowflake.Data.Tests/UnitTests/SFDataConverterTest.cs (hunk 1 above)
Comment (sfc-gh-dstempniak) [reply Snowflake.Data.Tests/UnitTests/SFDataConverterTest.cs:347]: good catch!

File: Snowflake.Data/Core/SFBindUploader.cs [hunk 4]
```diff
@@ -259,18 +259,30 @@ internal string GetCSVData(string sType, string sValue)
...
+
+                    DateTime ntz = epoch.AddTicks(ticksFromEpochNtz);
                     return ntz.ToString("yyyy-MM-dd HH:mm:ss.fffffff");
                 case "TIMESTAMP_TZ":
                     string[] tstzString = sValue.Split(' ');
-                    long nsFromEpochTz = long.Parse(tstzString[0]); // SFDateConverter provides in [ns] from Epoch
+                    long ticksFromEpochTz =
+                        long.TryParse(tstzString[0], out var nsTz) ?
+                            nsTz / 100 :
+                            (long)(decimal.Parse(tstzString[0]) / 100);
```
[the remark this answers is not in the document]
Comment (sfc-gh-mhofman) [reply Snowflake.Data/Core/SFBindUploader.cs:282]: nvm - offset is calculated from 2nd part of the input

**Review by sfc-gh-mhofman — APPROVED** [review]
besides this one what sfc-gh-knozderko found looks legit

File: Snowflake.Data.Tests/IntegrationTests/SFBindTestIT.cs (hunk 3 above)
Comment (sfc-gh-dstempniak) [reply Snowflake.Data.Tests/IntegrationTests/SFBindTestIT.cs:4]: added `restore`

File: Snowflake.Data/Core/SFBindUploader.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sfc-gh-dstempniak) [reply Snowflake.Data/Core/SFBindUploader.cs]: Yes, we divide on `decimal` and store the result on `long`.

File: Snowflake.Data.Tests/UnitTests/SFDataConverterTest.cs (hunk 2 above)
Comment (sfc-gh-dstempniak) [reply Snowflake.Data.Tests/UnitTests/SFDataConverterTest.cs:350]: added

File: Snowflake.Data/Core/SFDataConverter.cs [hunk 5]
```diff
@@ -325,24 +325,14 @@ internal static byte[] HexToBytes(string hex)
 
         internal static string csharpValToSfVal(SFDataType sfDataType, object srcVal)
         {
+            string TicksToNanoSecondsString(long tickDiff) => tickDiff == 0 ? "0" : $"{tickDiff}00";
```
[the remark this answers is not in the document]
Comment (sfc-gh-dstempniak) [reply Snowflake.Data/Core/SFDataConverter.cs]: Done. I've also renamed other methods from this file to be in line with our naming conventions.

File: Snowflake.Data.Tests/IntegrationTests/StructuredTypesWithEmbeddedUnstructuredIT.cs [hunk 6]
```diff
@@ -361,6 +361,41 @@ internal static IEnumerable<object[]> DateTimeConversionCases()
...
+            {
+                "9999-12-31 23:59:59.999999 +1:00",
+                SFDataType.TIMESTAMP_TZ.ToString(),
+                null,
+                DateTime.SpecifyKind(DateTime.Parse("9999-12-31 22:59:59.999999"), DateTimeKind.Utc)
+            };
+            yield return new object[]
+            {
+                "9999-12-31 23:59:59.999999 +13:00",
+                SFDataType.TIMESTAMP_LTZ.ToString(),
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data.Tests/IntegrationTests/StructuredTypesWithEmbeddedUnstructuredIT.cs:381]: maybe we could add timestamp_ltz for early years as well?

File: Snowflake.Data.Tests/IntegrationTests/StructuredTypesWithEmbeddedUnstructuredIT.cs [hunk 7]
```diff
@@ -445,6 +480,41 @@ internal static IEnumerable<object[]> DateTimeOffsetConversionCases()
...
+            {
+                "9999-12-31 23:59:59.999999 +1:00",
+                SFDataType.TIMESTAMP_TZ.ToString(),
+                null,
+                DateTimeOffset.Parse("9999-12-31 23:59:59.999999 +1:00")
+            };
+            yield return new object[]
+            {
+                "9999-12-31 23:59:59.999999 +13:00",
+                SFDataType.TIMESTAMP_LTZ.ToString(),
```
Comment (sfc-gh-knozderko) [inline Snowflake.Data.Tests/IntegrationTests/StructuredTypesWithEmbeddedUnstructuredIT.cs:500]: maybe we could add timestamp_ltz for early years as well?

**Review by sfc-gh-knozderko — APPROVED** [review]

File: Snowflake.Data.Tests/IntegrationTests/StructuredTypesWithEmbeddedUnstructuredIT.cs (hunk 6 above)
Comment (sfc-gh-dstempniak) [reply Snowflake.Data.Tests/IntegrationTests/StructuredTypesWithEmbeddedUnstructuredIT.cs:381]: There is a problem with ToLocalTime() with small dates like `0001-01-01`. It behaves differently on Linux and Windows. On Linux, .NET includes the historical data; on Windows, it does not. That's why I skipped such tests. Check this: https://developercommunity.visualstudio.com/t/Result-from-DateTimeToLocalTime-is-of/10576042

**Pull request merged**